src/AI/CoreLogic.pm

```diff
@@ -1777,48 +1779,44 @@ sub processAutoCart {
 		if (timeOut($AI::Timeouts::autoCart, $timeout) && $char->cartActive) {
 			my @addItems;
 			my @getItems;
-			my $cartInventory = $cart{inventory};
 			my $max;
 
-			if ($config{cartMaxWeight} && $cart{weight} < $config{cartMaxWeight}) {
-				foreach my $item (@{$char->inventory->getItems()}) {
-					next if ($item->{broken} && $item->{type} == 7); # dont auto-cart add pet eggs in use
-					next if ($item->{equipped});
-					my $control = items_control($item->{name});
-					if ($control->{cart_add} && $item->{amount} > $control->{keep}) {
+			if ($config{cartMaxWeight} && $char->cart->{weight} < $config{cartMaxWeight}) {
+				foreach my $invItem (@{$char->inventory->getItems()}) {
+					next if ($invItem->{broken} && $invItem->{type} == 7); # dont auto-cart add pet eggs in use
+					next if ($invItem->{equipped});
+					my $control = items_control($invItem->{name});
+					if ($control->{cart_add} && $invItem->{amount} > $control->{keep}) {
 						my %obj;
-						$obj{index} = $item->{invIndex};
-						$obj{amount} = $item->{amount} - $control->{keep};
+						$obj{index} = $invItem->{index};
+						$obj{amount} = $invItem->{amount} - $control->{keep};
 						push @addItems, \%obj;
-						debug "Scheduling $item->{name} ($item->{invIndex}) x $obj{amount} for adding to cart\n", "ai_autoCart";
+						debug "Scheduling $invItem->{name} ($invItem->{invIndex}) x $obj{amount} for adding to cart\n", "ai_autoCart";
 					}
 				}
 				cartAdd(\@addItems);
 			}
 
-			$max = @{$cartInventory};
-			for (my $i = 0; $i < $max; $i++) {
-				my $cartItem = $cartInventory->[$i];
-				next unless ($cartItem);
+			foreach my $cartItem (@{$char->cart->getItems()}) {
 				my $control = items_control($cartItem->{name});
 				next unless ($control->{cart_get});
 
-				my $item = $char->inventory->getByName($cartItem->{name});
+				my $invItem = $char->inventory->getByName($cartItem->{name});
 				my $amount;
-				if (!$item) {
+				if (!$invItem) {
 					$amount = $control->{keep};
-				} elsif ($item->{amount} < $control->{keep}) {
-					$amount = $control->{keep} - $item->{amount};
+				} elsif ($invItem->{amount} < $control->{keep}) {
+					$amount = $control->{keep} - $invItem->{amount};
 				}
 				if ($amount > $cartItem->{amount}) {
 					$amount = $cartItem->{amount};
 				}
 				if ($amount > 0) {
 					my %obj;
-					$obj{index} = $i;
+					$obj{index} = $cartItem->{index};
```

Comment:
I haven't looked over this whole pull request yet, but it seems like this is a breaking change. `AI::cartGet` expects to get the client index, not the server index. And while you've changed the implementation of `processCartGet` to match, so it doesn't actually break this particular use of `cartGet`, any other code which uses `cartGet` will break. Is there a benefit to using the server index? Does this protect us when we get disconnected in the middle of the sequence or something, for instance? (Since I know that will re-order the client indices.)